packages/core/src/service/fetcher.ts

```diff
@@ -0,0 +1,50 @@
+import { Dev } from '../common/index.js'
+import type { Externals, Logger } from '../common/index.js'
+
+export async function fetchWithCache(
+	{ web }: Externals,
+	logger: Logger,
+	input: RequestInfo | URL,
+	init?: RequestInit,
+): Promise<Response> {
+	const cache = await web.getCache()
+	const request = new Request(input, init)
+	const cachedResponse = await cache.match(request)
+	const cachedEtag = cachedResponse?.headers.get('ETag')
+	if (cachedEtag) {
+		request.headers.set('If-None-Match', cachedEtag)
+	}
+	try {
+		const response = await web.fetch(request, {
+			headers: { 'User-Agent': 'SpyglassMC (+https://spyglassmc.com)' },
+		})
+		if (response.status === 304) {
+			Dev.assertDefined(cachedResponse)
+			logger.info(`[fetchWithCache] reusing cache for ${request.url}`)
+			return cachedResponse
+		} else if (!response.ok) {
+			let message = response.statusText
+			try {
+				message = (await response.json()).message
+			} catch (e) {}
+			throw new TypeError(`${response.status} ${message}`)
+		} else {
+			try {
+				await cache.put(request, response)
```

Comment:
I noticed an error here when using this code in the browser. Over there it actually consumes the response body when calling `cache.put`, without cloning. So instead of putting `response.clone()` inside `HttpCache`, it should be put in this function.